File: 백준/Gold/1987. 알파벳/알파벳.py
import sys


# dfs
def dfs(x, y, z):
    global cnt

    cnt = max(cnt, z)

    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if 0 <= nx < r and 0 <= ny < c and visited[graph[nx][ny]] == 0:
            visited[graph[nx][ny]] = 1
            dfs(nx, ny, z + 1)
            visited[graph[nx][ny]] = 0

    return cnt

r, c = map(int, sys.stdin.readline().split())
graph = [list(map(lambda x: ord(x)-65, sys.stdin.readline().rstrip())) for _ in range(r)]
visited = [0] * 26 

dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]
cnt = 1
visited[graph[0][0]] = 1
print(dfs(0, 0, cnt))

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.*;
import java.io.IOException;
import java.util.*;

class Main {
    static int n, m;
    static int a, b, c, d, answer;
    static int[] visited;
    static char[][] arr;
    static int[] dx = { 1, -1, 0, 0 };
    static int[] dy = { 0, 0, -1, 1 };

    // static LinkedList<Integer> result_dfs = new LinkedList<Integer>();
    // static LinkedList<Integer> result_bfs = new LinkedList<Integer>();

    public static boolean is_move(int x, int y) {
        if (0 <= x && x < n && 0 <= y && y < m) {
            return true;
        }
        return false;
    }

    public static void dfs(int x, int y, int cnt) {
        answer = Math.max(answer, cnt);
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (is_move(nx, ny) && visited[(int) arr[nx][ny]] == 0) {
                visited[(int) arr[nx][ny]] = 1;
                dfs(nx, ny, cnt + 1);
                visited[(int) arr[nx][ny]] = 0;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;
        answer = 0;
        visited = new int[200];
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        arr = new char[n][m];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            String a = st.nextToken();
            for (int j = 0; j < m; j++) {
                arr[i][j] = a.charAt(j);
            }
        }
        visited[arr[0][0]] = 1;
        dfs(0, 0, 1);
        System.out.println(answer);
        br.close();
        bw.close();
    }

}